### PR TITLE
Finalize forecast trust and replay calibration research

### PR DESCRIPTION
**Pull request by xmarre**
## Final state

This PR closes the forecast-trust/replay-calibration research cycle and retains only the mechanisms that survived the completed gates.

### Supported

- **Generic causal latest-delta correction (PR #39): retained.** It remains the correction applied by `model_aware_mode=full` on the causal single-pass path and adds no transformer evaluation.
- **`full` for current native ER-SDE quality testing:** among the compared model-aware configurations, `full` single-pass is the currently preferred tested mode. This conclusion is scoped to native ER-SDE under the tested H3 setup.

### Supported alternative

- **Offline smoothing replay remains supported and remains the compatibility-safe global default.** Its historical MiniMax H3 audio/stutter rationale still stands. Current native ER-SDE testing found a temporal facial case where `full` single-pass performed better than the tested replay path; that does not establish the same ranking for Euler, RES/RES CFG++, Turbo/LightX2V, or other samplers.

### Default off / not recommended for promotion

```text
model_aware_trust_shrinkage = false
```

Causal trust shrinkage produced large hidden-feature observer improvements, but the completed real perceptual A/B gate did not show a reliable user-facing gain and sometimes slightly favored trust disabled. The gate is complete. No denser kappa search, rescue value, or replacement trust formula is introduced.

The switch remains for research/reproduction and saved-workflow compatibility.

### Rejected transfers

- causal trust kappa -> future-bracket replay geometry;
- causal PR #39 latest-delta scalar -> future-bracket replay direction.

Accordingly:

```text
model_aware_replay_generic_correction = false
```

remains the supported default. `true` is retained only for legacy/scientific reproduction.

### Research infrastructure only

The PR retains:

- exact VIDEO replay-calibration export;
- exact quadratic moments and ratio-normalization parity;
- source/config/schedule/topology/trace provenance and fingerprints;
- structural validation of interior withheld targets;
- CPU-only offline evaluator;
- whole-run validation;
- Level 0 fixed controls;
- Level 1 affine current-weight baseline;
- one-predictor Level 2 residual tests with coordinate control.

No applied affine, disagreement, validation-penalty, coordinate, floor, interaction, tree, neural, AutoML, or other replay controller is shipped.

### Closed hypotheses

The following are closed as complete runtime solutions for this cycle:

- pure global alpha;
- K=2;
- FinalLayer-transformed correction;
- previous-error direction families;
- hold-anchor replay shrinkage;
- causal kappa -> replay transfer;
- causal generic scalar -> replay transfer.

## Final controlled native ER-SDE gate

All comparison runs below used the same seed, prompt, starting image, resolution, duration/frame count, native ER-SDE sampler/scheduler setup, model, precision, workflow, and remaining generation settings.

### `schedule_confidence`, single pass

```text
steps = 20
actual_steps = 11
forecast_steps = 9
actual_transformer_calls = 11
fallbacks = 0
model_aware_extra_nfes = 0

audio hidden forecast ratio = 1.789477
video hidden forecast ratio = 1.309651
```

The recurring false eye-motion artifact remained subtly visible immediately before the intended eye closure.

### `full`, single pass

The expensive schedule was identical:

```text
actual_steps = 11
forecast_steps = 9
actual_transformer_calls = 11
fallbacks = 0
model_aware_extra_nfes = 0
```

The meaningful added mechanism is the retained generic latest-delta causal correction:

```text
audio: 1.777636 -> 1.670690   (~6.02% hidden-error reduction)
video: 1.313055 -> 1.250087   (~4.80% hidden-error reduction)

model_aware_correction_max = 0.148827
model_aware_causal_correction_s = 0.000381
model_aware_extra_nfes = 0
```

Representative later video anchors improved consistently:

```text
1.233778 -> 1.166099
1.278252 -> 1.200908
1.391828 -> 1.307558
1.392824 -> 1.316612
1.153456 -> 1.103306
```

Perceptually, the recurring false eye-motion artifact was absent in `full`; the eyes remained synchronized with the intended closure.

The `6.02%` and `4.80%` figures are hidden-feature-error reductions on this trace. They are not perceptual-quality percentages.

### Offline smoothing replay

```text
model_aware_mode = off
offline_smoothing_replay = true

first pass: 11 actual / 9 forecast / 11 transformer calls
replay: 20 steps / 11 anchors / 9 smoothed / 0 transformer calls
replay wall time ~= 13.25 s
archive ~= 4315.5 MiB
effective video blend mean ~= 0.346053
audio = local-only
```

The abnormal eye motion remained visibly present and was worse than `schedule_confidence` for this artifact and clearly worse than `full`.

This result is **native-ER-SDE scoped**. Offline replay remains supported because its earlier H3 audio/stutter evidence is still valid.

Plain/base Spectrum was not part of this exact three-run eye-motion comparison, so no base-Spectrum result is claimed here.

## Supporting native ER-SDE perceptual evidence

An earlier controlled same-seed ER-SDE comparison containing:

```text
meine Freunde sind noch 2 Stunden weg
```

isolated a pronunciation difference: the base/schedule-confidence-like path pronounced `weg` incorrectly, while `full` retained the intended pronunciation. A direct follow-up found `schedule_confidence` sounding like the base path and `full` retaining the correction. Offline replay also produced the intended pronunciation in the earlier comparison.

This is supporting ER-SDE evidence for the generic causal correction, not a universal speech-quality theorem.

## ER-SDE terminal replay policy correction

The previous v0.2.7 fix over-generalized the reproduced 25-step failure into a blanket `min_tail_actual_steps >= 2` rule for all ER-SDE runs.

The actual failure geometry was:

```text
22 actual
23 forecast
24 actual
```

During **offline capture only**, the runtime now promotes the penultimate ER-SDE step **only when the normal schedule would otherwise forecast it**. The reproduced 25-step terminal pattern becomes `22/23/24 actual`.

Normal 20-step and 32-step ER-SDE schedules already make their penultimate step actual, so they receive no additional forced transformer evaluation. Explicitly larger configured tails still win. RES keeps its independent three-step tail policy. Euler/RES/Turbo policies are unchanged.

## Native ER-SDE replay compatibility

This branch incorporates merged PR #43. Native/default ER-SDE seeded replay accepts the reviewed deterministic native `noise_scaler` closures while continuing to reject explicit custom `noise_sampler` instances, arbitrary/custom/stateful scalers, and unknown future closure contracts.

## Defaults and compatibility

Shipping defaults remain compatibility-safe:

```text
model_aware_mode = off
model_aware_trust_shrinkage = false
model_aware_replay_generic_correction = false
offline_smoothing_replay = true
```

For **current native ER-SDE quality testing**, the preferred tested configuration is documented as:

```text
model_aware_mode = full
model_aware_trust_shrinkage = false
offline_smoothing_replay = false
```

No global default was flipped from ER-SDE-only evidence. Existing saved input values continue to be honored.

## Release preparation

The branch is prepared for `v0.2.8` using the repository's existing tested-main release workflow, with updated README, benchmark record, package version, and release notes. Final merge/release remains gated on current-head multi-revision CI and final review-thread resolution.

### CODE REVIEW

**Comment by xmarre** [comment]
@coderabbitai review

**Comment by xmarre** [comment]
25-step real H3 trace recorded in `FORECAST_TRUST_BENCHMARK.md` (commit `6d5fde4`). This was not the prescribed 20-step parity run, so I am not treating it as the exact PR-39 parity gate. It is strong mechanism evidence for the 25-step regime.

Key first-pass results (12 audio + 12 video samples, zero probe failures, zero probe-added transformer NFE):

- current generic-corrected ratio: audio `1.542742`, video `1.240801`
- bounded same-delta advantage: audio `+3.6977%`, video `+3.7016%`
- trust oracle advantage: audio `+34.9660%`, video `+19.3114%`
- oracle kappa: audio `0.089179`, video `0.035404`
- disagreement/error correlation: audio `0.859295`, video `0.559713`
- disagreement/required-shrink correlation: audio `0.907817`, video `0.751129`
- best fixed candidate in the current sweep was `theta=0.15`: audio ratio `0.990551` (`+34.7545%`), video ratio `1.005021` (`+18.7989%`)

This clearly selects the trust-distance branch over further scalar-gain tuning for this regime: its measured headroom is ~9.5x larger on audio and ~5.2x larger on video.

Important limitation before applying it: the ordinary trace is horizon-2-heavy (`horizon_mean=1.8333`, max 2), while the live ER-SDE forecast decisions in this run are horizon 1. The next gate therefore keeps the same 25-step configuration and sets `model_aware_risk_threshold=0.0` so would-be forecast locations become exact evaluations. That gives direct horizon-1 ground truth for the fixed theta sweep.

If the horizon-1 trace stays positive on both streams, the next patch will add an opt-in causal first-pass trust controller for same-seed 25-step A/B. Offline smoother weights will remain unchanged in that first applied experiment so the result isolates better causal anchors instead of combining two mechanisms. The good 20-step regime will then be rerun as the regression gate before considering any default change.

**Comment by xmarre** [comment]
@coderabbitai review

File: comfyui_spectrum_h3/trust_probe.py [hunk 1]
```diff
@@ -0,0 +1,1138 @@
...
+            blend_weight=float(blend),
+            correction_gain=float(gain),
+            disagreement=float(disagreement),
+            kappa=trust_kappa(
+                disagreement,
+                horizon,
+                theta=_APPLIED_THETA,
+            ),
+        )
+    )
```
[the remark this answers is not in the document]
Comment (xmarre) [reply comfyui_spectrum_h3/trust_probe.py:680]: @coderabbitai Addressed. `_ReplayShadowRecord.kappa` is now explicitly documented as an anchor-step shadow/probe coefficient, distinct from the applied forecast-step kappa. The comment also states that `model_aware_trust_replay_shadow_*` is diagnostic and must not be interpreted as applied-path evidence. Implemented in `e83cfeff`.

File: comfyui_spectrum_h3/trust_probe.py [hunk 2]
```diff
@@ -0,0 +1,1138 @@
...
+    if not isinstance(records, list) or not records:
+        return
+    ranges = {name: (start, end) for name, start, end in smoother._stream_ranges}
+    samples_by_stream = {
+        name: _sample_archive_stream(smoother, start, end)
+        for name, (start, end) in ranges.items()
+        if name in {"audio", "video"}
+    }
+    anchor_ids = list(smoother._anchor_ids)
```
[the remark this answers is not in the document]
Comment (xmarre) [reply comfyui_spectrum_h3/trust_probe.py]: @coderabbitai Addressed. The replay-shadow setup is now isolated from the applied replay build: ordinary setup failures increment `replay_shadow_failures` and return after the trust weights have already been built, while `torch.cuda.OutOfMemoryError` still propagates. Added regressions for both non-OOM diagnostic failure and OOM propagation. Implemented in `e83cfeff` / `be1c4485`.

File: tests/test_trust_probe.py [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+                kappa=audio_kappa,
+            ),
+            video=trust_module._StreamTrustDecision(
+                disagreement=0.3,
+                kappa=video_kappa,
+            ),
+            compute_seconds=0.0,
+            scalar_transfer_seconds=0.0,
+        )
+    }
```
[the remark this answers is not in the document]
Comment (xmarre) [reply tests/test_trust_probe.py:139]: @coderabbitai Addressed. Added an offline first-pass persistence integration test that exercises the installed `_prediction_segments` → `_ensure_step_trust` path instead of directly injecting `_model_aware_trust_forecasts`. It verifies the persisted `step_id` and `latest_anchor_id`, constructs `OfflineSmoother`, then confirms the matching replay step changes while a different forecast step remains baseline-identical. Implemented in `be1c4485`.

File: tests/test_trust_probe.py [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+    run_id = runtime.start_run(
+        sigmas,
+        "sample_er_sde",
+        supported_sampler=True,
+    )
+    assert runtime._run is not None
+    assert runtime._run.min_tail_actual_steps >= 2
+    summary = runtime.debug_summary()
+    assert "model_aware_trust_extra_transformer_nfe=0" in summary
+    runtime.end_run(run_id)
```
[the remark this answers is not in the document]
Comment (xmarre) [reply tests/test_trust_probe.py]: @coderabbitai Addressed. Replaced the metadata-only ER-SDE assertion with a counted six-step offline first-pass execution for trust disabled/enabled. The regression compares the actual/forecast decision sequence and `actual_transformer_calls`, asserts zero fallbacks, verifies the required two-logical-step exact tail, and confirms the enabled path actually records trust applications without failures. Implemented in `be1c4485`.

**Comment by xmarre** [comment]
@coderabbitai Nitpick (`trust_probe.py`, shared weight-fit cost) addressed. `_spectral_weights_configured` and `_linear_weights` are now computed once per trust evaluation and reused for both audio and video. Added a regression test that instruments both methods and asserts exactly one call each. Implemented in `e83cfeff` / `be1c4485`.

**Comment by xmarre** [comment]
@coderabbitai Nitpick (`model_aware_trust_shrinkage` mode compatibility) addressed. `SpectrumH3Config.validate()` now rejects `model_aware_trust_shrinkage=True` unless `model_aware_mode='full'`, so the opt-in can no longer silently no-op in `schedule` / `schedule_confidence` / `off`. Added a regression assertion for the incompatible combination. Implemented in `481ed253` / `be1c4485`.

**Comment by xmarre** [comment]
@coderabbitai review

**Comment by xmarre** [comment]
@coderabbitai Nitpick (`tests/test_trust_probe.py`, duplicated replay-shadow record fixture) addressed. Extracted a single `_shadow_record()` helper beside the existing test helpers and replaced both identical `_ReplayShadowRecord` literals, leaving the two failure-injection paths unchanged. Implemented in `732fc0fa`.

**Comment by xmarre** [comment]
@coderabbitai review

**Comment by xmarre** [comment]
@coderabbitai review

File: comfyui_spectrum_h3/replay_trust_shadow.py [hunk 5]
```diff
@@ -0,0 +1,514 @@
...
+                    else aggregate.replay_shadow_video
+                )
+                legacy.record(
+                    case["baseline_ratio"],
+                    case["causal_transfer_ratio"],
+                )
+            except torch.cuda.OutOfMemoryError:
+                raise
+            except (RuntimeError, TypeError, ValueError, KeyError, IndexError):
+                aggregate.replay_shadow_failures += 1
```
[the remark this answers is not in the document]
Comment (xmarre) [reply comfyui_spectrum_h3/replay_trust_shadow.py:563]: @coderabbitai Addressed. This was valid: the per-record diagnostic handler now catches `AttributeError` alongside the other ordinary diagnostic exceptions, while `torch.cuda.OutOfMemoryError` still re-raises. Ordinary replay-shadow failures therefore increment `replay_shadow_failures` without aborting the already-built replay weights. Implemented in `e73fcdb0`.

File: comfyui_spectrum_h3/replay_trust_shadow.py [hunk 6]
```diff
@@ -0,0 +1,514 @@
...
+        f"model_aware_trust_replay_shadow_{name}_causal_transfer_advantage_mean={stream.mean(stream.causal_transfer_advantage_sum):.6f}",
+        f"model_aware_trust_replay_shadow_{name}_causal_disagreement_error_corr={stream.causal_error_correlation.correlation():.6f}",
+        f"model_aware_trust_replay_shadow_{name}_causal_disagreement_shrink_corr={stream.causal_shrink_correlation.correlation():.6f}",
+        f"model_aware_trust_replay_shadow_{name}_observer={observer_mode}",
+        f"model_aware_trust_replay_shadow_{name}_observer_samples={stream.replay_observer_count}",
+        f"model_aware_trust_replay_shadow_{name}_replay_disagreement_mean={stream.observer_mean(stream.replay_disagreement_sum):.6f}",
+        f"model_aware_trust_replay_shadow_{name}_replay_disagreement_max={stream.replay_disagreement_max:.6f}",
+        f"model_aware_trust_replay_shadow_{name}_replay_disagreement_error_corr={stream.replay_error_correlation.correlation():.6f}",
+        f"model_aware_trust_replay_shadow_{name}_replay_disagreement_shrink_corr={stream.replay_shrink_correlation.correlation():.6f}",
+    ]
```
[the remark this answers is not in the document]
Comment (xmarre) [reply comfyui_spectrum_h3/replay_trust_shadow.py:621]: @coderabbitai I checked this against the composed debug-summary path and did not add the proposed field a second time. `model_aware_trust_replay_shadow_{stream}_baseline_ratio_mean` is already emitted by the wrapped trust summary (`_replay_shadow_summary`), and the replay-native validator deliberately populates that legacy accumulator with the same validation-attenuated replay baseline via `legacy.record(case["baseline_ratio"], ...)`. Adding it again in `_native_stream_summary` would produce a duplicate telemetry key. I added an explicit code comment documenting that ownership in `e73fcdb0`. The existing baseline field therefore remains emitted once and corresponds to the replay-native baseline used by the appended oracle/advantage fields.

**Comment by xmarre** [comment]
@coderabbitai review

**Comment by xmarre** [comment]
@coderabbitai review

**Comment by xmarre** [comment]
@coderabbitai review

**Comment by xmarre** [comment]
@coderabbitai Addressed the replay-gate telemetry nitpick. Enabled-path `model_aware_replay_generic_correction_applications` now comes from the smoother's observed `model_aware_offline_correction_applications` counter after the replay build rather than the static opportunity count. The opportunity count is retained only for disabled-path skips. Implemented in `499aa073`.

**Comment by xmarre** [comment]
@coderabbitai Addressed the positional-cursor maintainability nitpick. `_decomposition_case` now serializes an ordered list of named tensor entries, transfers the stacked values once, then reconstructs results by those names. Candidate, scalar, axis, and optional-field semantics are unchanged, but later insertions can no longer silently shift unrelated fields. Implemented in `9f6c76bc`.

**Comment by xmarre** [comment]
@coderabbitai Addressed the unreachable-wrapper nitpick. The package-supported path always installs replay-shadow composition immediately after component registration, so the intermediate `_validate_replay_native_shadow_with_decomposition` hook was dead in the shipped configuration. It has been removed; component installation now captures the native validator for the composition layer and installs only its debug-summary state. Implemented in `9f6c76bc`.

**Comment by xmarre** [comment]
@coderabbitai Addressed the composed-validator coverage nitpick. Added a focused regression that supplies actual replay-shadow records, injects one recoverable native diagnostic failure, invokes `_validate_composed_replay_shadows`, and verifies the failure count remains exactly one while both component stream counts remain zero. The new test is also included in Ruff/compileall CI. Implemented in `0c1d636a` / `e927a2e5`.

**Comment by xmarre** [comment]
@coderabbitai review

File: tests/test_replay_shadow_composition.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        composition_module,
+        "_NATIVE_REPLAY_VALIDATOR",
+        recoverable_native_failure,
+    )
+    composition_module._validate_composed_replay_shadows(smoother, aggregate)
+
+    component = component_module._component_aggregate(archive)
+    assert aggregate.replay_shadow_failures == 1
+    assert component.audio.count == 0
+    assert component.video.count == 0
```
[the remark this answers is not in the document]
Comment (xmarre) [reply tests/test_replay_shadow_composition.py:67]: @coderabbitai Addressed. The composed-validator regression now spies directly on `component_module._validate_replay_decomposition` and asserts the call count remains exactly zero after the recoverable native failure, while retaining the single-failure and zero-stream-count assertions. Implemented in `1095fd05`.

**Comment by xmarre** [comment]
@coderabbitai review

**Comment by xmarre** [comment]
@coderabbitai review

File: FORECAST_TRUST_BENCHMARK.md [hunk 8]
```diff
@@ -0,0 +1,593 @@
...
+```text
+audio B advantage vs D   +7.8915%
+video B advantage vs D   +5.2181%
+```
+
+The replay-geometry diagnosis remains supported.
+
+### Perceptual result
+
+The same-seed generated D and B outputs were judged **indistinguishable** by the user:
```
[the remark this answers is not in the document]
Comment (xmarre) [reply FORECAST_TRUST_BENCHMARK.md]: @coderabbitai Addressed. The benchmark now states directly that the D and B outputs were generated with the same seed while preserving the authoritative indistinguishable/neutral perceptual result. Implemented in `26168619`.

**Comment by xmarre** [comment]
@coderabbitai review

**Comment by xmarre** [comment]
@coderabbitai review

**Comment by xmarre** [comment]
@coderabbitai review

**Comment by xmarre** [comment]
@coderabbitai review

File: comfyui_spectrum_h3/replay_calibration.py [hunk 9]
```diff
@@ -0,0 +1,704 @@
...
+        candidates=candidates,
+        effective_blends=effective_blends,
+    )
+
+    left = smoother.archive.anchors[target_index - 1]
+    right = smoother.archive.anchors[target_index + 1]
+    spacing = float(right.coordinate - left.coordinate)
+    if abs(spacing) <= 1e-12:
+        raise RuntimeError("replay calibration bracket has duplicate coordinates")
+    bracket_fraction = (float(record.coordinate) - float(left.coordinate)) / spacing
```
[the remark this answers is not in the document]
Comment (xmarre) [reply comfyui_spectrum_h3/replay_calibration.py:259]: Fixed. The structural guard module already existed, but the production package installer was not activating it. The final branch now imports and installs `install_replay_calibration_validation()` after replay calibration and before provenance. The guard rejects any VIDEO calibration target that is not strictly interior before left/right anchor indexing or blend evaluation, while preserving the existing diagnostic-failure isolation for valid rows. Existing `tests/test_replay_calibration_validation.py` covers the boundary cases.

File: README.md [hunk 10]
```diff
@@ -509,18 +532,26 @@ ComfyUI-Spectrum-MiniMax-H3/
...
+|   |-- replay_generic_correction_gate.py
+|   |-- replay_spectral_alpha_shadow.py
+|   |-- replay_spectral_mixture_shadow.py
+|   |-- replay_trust_shadow.py
 |   |-- runtime.py
 |   |-- rollback.py
 |   |-- sampling.py
 |   `-- minimax_h3.py
+|-- tools/
+|   `-- analyze_replay_calibration.py
```
[the remark this answers is not in the document]
Comment (xmarre) [reply README.md:639]: Fixed. The final README repository tree now lists both `replay_component_shadow.py` and `replay_shadow_composition.py`, along with the other replay-calibration/validation and ER-SDE policy modules that are part of the shipping package.

File: tools/analyze_replay_calibration.py [hunk 11]
```diff
@@ -0,0 +1,825 @@
...
+            if not key or not value:
+                raise CalibrationError(f"invalid --{kind} annotation {spec!r}")
+            resolved[str(pathlib.Path(key))] = value
+        else:
+            if len(inputs) != 1:
+                raise CalibrationError(
+                    f"bare --{kind} requires exactly one input; use INPUT=VALUE with multiple inputs"
+                )
+            resolved[str(pathlib.Path(inputs[0]))] = spec
+    return resolved
```
[the remark this answers is not in the document]
Comment (xmarre) [reply tools/analyze_replay_calibration.py:267]: Fixed in the PR #46 review follow-up merged into this branch. `_parse_annotation_specs` now normalizes the supplied input set and rejects every explicit `INPUT=VALUE` annotation whose key matches no input, while preserving existing bare-spec, value-validation, and path-normalization behavior. This prevents a misspelled external seed/label annotation from silently weakening run identity or fold labeling.
